omisego/elixir-omg
# feat: use open-api generated client to run tests

## Overview

Use open-api generated client to run test. Also add a smoke test to make sure the client is working + to be helpful in general to checkout if services are up before checking other future tests failure.

## Changes

- Add `Makefile` to: 
    1. generate the client
    1. spin the services up/down
    1. test
- Have a script that would auto generate the open-api client by our swagger spec using the docker provided by open-api: [GH doc](https://github.com/OpenAPITools/openapi-generator#16---docker)
- Provide some custom `Connection` code to override the middleware to swap the service url to the one instead of the one provided by the auto generated client. Using this idea: [here](https://github.com/teamon/tesla#runtime-middleware)
- Add smoke test (replacing the dummy test) that hits one simple api of child chain, watcher security, and watcher info to check connection is correct and demo that we are using the generated client correctly.
- Replace the adapter of `Tesla` (request library of elixir) to `Hackney` due to the issue that default adapter cannot send request with empty value when the api is a `post` request: [link](https://github.com/googleapis/elixir-google-api/issues/26#issuecomment-360209019)
- Update README on how to setup the env var for overriding urls and run the test

## Testing

```
cd priv/perf
make start-services
make init
make test
make stop-services
```
